util/promlint/promlint.go

```diff
@@ -0,0 +1,208 @@
+// names, types, and metadata, and reports them to the caller.
+type Linter struct {
+	r io.Reader
+}
+
+// A Problem is an issue detected by a Linter.
+type Problem struct {
+	// The name of the metric indicated by this Problem.
+	Metric string
+
+	// A description of the issue for this Problem.
+	Text string
+}
+
+// New creates a new Linter that reads an input stream of Prometheus metrics.
+// Only the text exposition format is supported.
+func New(r io.Reader) *Linter {
+	return &Linter{
+		r: r,
+	}
+}
+
+// Lint performs a linting pass, returning a slice of Problems indicating any
+// issues found in the metrics stream.
+func (l *Linter) Lint() ([]Problem, error) {
+	// TODO(mdlayher): support for protobuf exposition format?
+	d := expfmt.NewDecoder(l.r, expfmt.FmtText)
+
+	var problems []Problem
+
+	var mf dto.MetricFamily
+	for {
+		if err := d.Decode(&mf); err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return nil, err
+		}
+
+		problems = append(problems, lint(mf)...)
+	}
+
+	// Ensure deterministic output.
+	sort.SliceStable(problems, func(i, j int) bool {
+		if problems[i].Metric < problems[j].Metric {
+			return true
+		}
+
+		return problems[i].Text < problems[j].Text
+	})
+
+	return problems, nil
+}
+
+// lint is the entry point for linting a single metric.
+func lint(mf dto.MetricFamily) []Problem {
+	// TODO(mdlayher): lint rules for specific metrics types.
+	return lintGeneral(mf)
+}
+
+// lintGeneral detects general issues not tied to any specific metric type.
+func lintGeneral(mf dto.MetricFamily) []Problem {
+	var problems []Problem
+
+	// Expect all metrics to have help text available.
+	if mf.Help == nil {
+		problems = append(problems, Problem{
+			Metric: *mf.Name,
+			Text:   "no help text",
+		})
+	}
+
+	problems = append(problems, lintMetricUnits(mf)...)
+
+	return problems
+}
+
+// lintMetricUnits detects issues with metric unit names.
+func lintMetricUnits(mf dto.MetricFamily) []Problem {
+	var problems []Problem
+
+	unit, base, ok := metricUnits(*mf.Name)
+	if !ok {
+		// No known units detected.
+		return nil
+	}
+
+	// Special case: express data units in bytes instead of *bits.
+	// Bits should recommend conversion to bytes instead.
+	if base == unitBits {
+		problems = append(problems, Problem{
+			Metric: *mf.Name,
+			Text:   fmt.Sprintf(`use base unit "bytes" instead of %q`, unit),
+		})
+
+		return problems
+	}
+
+	// Unit is already a base unit.
+	if unit == base {
+		return nil
+	}
+
+	problems = append(problems, Problem{
+		Metric: *mf.Name,
+		Text:   fmt.Sprintf("use base unit %q instead of %q", base, unit),
+	})
+
+	return problems
+}
+
+// metricUnits attempts to detect known unit types used as the suffix of a
+// piece of a metric name, e.g. "foo_bytes" or "bar_baz_milligrams".
+// It returns the detected unit, the base unit (no prefix), and whether or
+// not a result was found.
+func metricUnits(m string) (string, string, bool) {
```

Comment:
I think that's one of the few cases where named return variables add clarity, especially in the docs. Would obsolete the sentence above as well.